### PR TITLE
Add all students from course to exam 

### PR DESCRIPTION
**Pull request by nicolasruscher**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple integration tests (Jest) related to the features (with a high test coverage)
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Currently adding multiple students to an exam can only be done via a csv file. Registering all students who are enrolled in the course is useful when creating exams. 

### Description

We add a Button the Instructor Page 'exam-students', which instructors can use the register all students to the exam which are currently enrolled in the course.  (E.g. the student has the courses student-group assigned)

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration -> Course -> Exam -> Students
3. Clicking on the new Button should register all students to the exam, who are currently enrolled in this course. 
4. This will enroll all users into the exam who are part of the courses student-group! 
5. No other students get registered to the exam. 
6. **Instructors of this course** or **Admins** will not be added to the exam! 

### Test Coverage

ExamResourse: 98.5%
exam-students-component: 79%
exam-managment-component: 85.7%
exam-managment-service: 81%

### Screenshots

Bildschirmfoto 2021-01-30 um 13 49 29

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/ExamService.java [hunk 1]
```diff
@@ -1046,4 +1043,25 @@ public void unregisterStudentFromExam(Long examId, boolean withParticipationsAnd
...
+    /**
+     * Adds all students registered in the course to the given exam
+     *
+     * @param courseId Id of the course
+     * @param examId Id of the exam
+     * @return the list of students who could not be registered for the exam, because they could NOT be found in the Artemis database and could NOT be found in the TUM LDAP
+     */
+    public List<StudentDTO> addAllStudentsOfCourseToExam(Long courseId, Long examId) {
+        Course course = courseService.findOne(courseId);
+        var users = userService.findAllUsersInGroup(course.getStudentGroupName());
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/ExamService.java:1056]: I think you can directly assign those users to the exam

**Review by airelawaleria — DISMISSED** [review]

File: src/main/java/de/tum/in/www1/artemis/service/ExamService.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ExamService.java:1058]: Looks better now. I suggest you use the method `getStudents(course)` here instead, which would be easier ;-)

image

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java [hunk 2]
```diff
@@ -601,6 +601,27 @@ public ExamResource(UserService userService, CourseService courseService, ExamSe
...
     }
 
+    /**
+     * POST /courses/:courseId/exams/:examId/addAllStudentsOfCourse : Add all users which are enrolled in the course to the exam so that the student can access the exam
+     *
+     * @param courseId     the id of the course
+     * @param examId       the id of the exam
+     * @return empty ResponseEntity with status 200 (OK) or with status 404 (Not Found)
+    */
+    @PostMapping(value = "/courses/{courseId}/exams/{examId}/addAllStudentsOfCourse")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:611]: I don't really like the last part `addAllStudentsOfCourse`
I guess `register-course-students` might be better

Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:611]:
```suggestion
    @PostMapping(value = "/courses/{courseId}/exams/{examId}/register-course-students")
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java [hunk 3]
```diff
@@ -601,6 +601,27 @@ public ExamResource(UserService userService, CourseService courseService, ExamSe
...
+    @PreAuthorize("hasAnyRole('INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Void> addAllStudentsOfCourseToExam(@PathVariable Long courseId, @PathVariable Long examId) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:613]:
```suggestion
    public ResponseEntity<Void> registerCourseStudents(@PathVariable Long courseId, @PathVariable Long examId) {
```

File: src/main/webapp/app/exam/manage/exam-management.service.ts [hunk 4]
```diff
@@ -148,6 +148,15 @@ export class ExamManagementService {
...
     }
 
+    /**
+     * Add all students of the course to the exam
+     * @param courseId
+     * @param examId
+     * @return studentDtos of students that were not found in the system
+     */
+    addAllStudentsOfCourseToExam(courseId: number, examId: number): Observable<HttpResponse<StudentDTO[]>> {
+        return this.http.post<any>(`${this.resourceUrl}/${courseId}/exams/${examId}/addAllStudentsOfCourse`, { observe: 'response' });
```
Comment (krusche) [inline src/main/webapp/app/exam/manage/exam-management.service.ts:158]:
```suggestion
        return this.http.post<any>(`${this.resourceUrl}/${courseId}/exams/${examId}/register-course-students`, { observe: 'response' });
```

File: src/main/webapp/i18n/de/exam.json [hunk 5]
```diff
@@ -247,6 +247,7 @@
                 "lastName": "Nachname",
                 "email": "E-Mail",
                 "registrationNumber": "Matrikelnummer",
+                "registerAllFromCourse": "Registriere alle zu diesem Kurs angemeldeten Studenten",
```
Comment (krusche) [inline src/main/webapp/i18n/de/exam.json:250]: This is too long for a button. I suggest to add a tooltip and make the button text shorter, e.g.

Comment (krusche) [reply src/main/webapp/i18n/de/exam.json:250]:
```suggestion
                "registerAllFromCourse": "Registriere Kurs Studierende",
```

File: src/main/webapp/i18n/en/exam.json [hunk 6]
```diff
@@ -248,6 +248,7 @@
                 "lastName": "Last name",
                 "email": "Email",
                 "registrationNumber": "Matriculation number",
+                "registerAllFromCourse": "Register all students currently enrolled for this course",
```
Comment (krusche) [inline src/main/webapp/i18n/en/exam.json:251]:
```suggestion
                "registerAllFromCourse": "Register course students",
```

File: src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java [hunk 7]
```diff
@@ -907,6 +909,39 @@ public void testGetExamForConduction() throws Exception {
...
+        var student99 = ModelFactory.generateActivatedUser("student99");     // not registered for the course
+        student99.setRegistrationNumber("1234");
+        userRepo.save(student99);
+        student99 = userRepo.findOneWithGroupsAndAuthoritiesByLogin("student99").get();
+        student99.setGroups(Collections.singleton("tumuser"));
+        userRepo.save(student99);
+        assertThat(student99.getGroups()).contains(course.getStudentGroupName());
+        assertThat(exam.getRegisteredUsers()).doesNotContain(student99);
+
+        request.postWithoutLocation("/api/courses/" + course.getId() + "/exams/" + exam.getId() + "/addAllStudentsOfCourse", null, HttpStatus.OK, null);
```
Comment (krusche) [inline src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java:935]:
```suggestion
        request.postWithoutLocation("/api/courses/" + course.getId() + "/exams/" + exam.getId() + "/register-course-students", null, HttpStatus.OK, null);
```

**Review by krusche — CHANGES_REQUESTED** [review]
Thanks for the contribution 👍 The code looks good, except some smaller issues, see my inline comments

**Review by JohannesWeiss — APPROVED** [review]
Tested on ts1, works as expected.

**Review by DenizShabani — DISMISSED** [review]
Works as expected on TS1

**Review by simonzilker — APPROVED** [review]
Works fine!

**Review by philgabr — COMMENTED** [review]
Works as described :+1:

One minor remark:
with the button, it is possible to register instructors/admins to the exam, which is not possible by adding the students one by one:
image

**Review by marvinchimeka — APPROVED** [review]
nice feature

**Review by ivanchimeno — CHANGES_REQUESTED** [review]
Nice feature :) I didn't find anything but I can't approve because there are requested changes

**Review by sebastiantpeter — APPROVED** [review]

**Review by valeriebucher — DISMISSED** [review]

**Review by EneaGore — COMMENTED** [review]
Functionality wise it seems fine, but it also adds enrolled intrusctors.

File: src/main/java/de/tum/in/www1/artemis/service/ExamService.java (hunk 1 above)
Comment (nicolasruscher) [reply src/main/java/de/tum/in/www1/artemis/service/ExamService.java]: Done!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java (hunk 2 above)
Comment (nicolasruscher) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java]: done!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java (hunk 3 above)
Comment (nicolasruscher) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ExamResource.java:613]: done!

File: src/main/webapp/app/exam/manage/exam-management.service.ts (hunk 4 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exam/manage/exam-management.service.ts]: done!

File: src/main/webapp/i18n/de/exam.json (hunk 5 above)
Comment (nicolasruscher) [reply src/main/webapp/i18n/de/exam.json]: done!

File: src/main/webapp/i18n/en/exam.json (hunk 6 above)
Comment (nicolasruscher) [reply src/main/webapp/i18n/en/exam.json]: done!

File: src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java (hunk 7 above)
Comment (nicolasruscher) [reply src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java]: done!

Comment (nicolasruscher) [reply src/test/java/de/tum/in/www1/artemis/ExamIntegrationTest.java]: done!

File: src/main/webapp/i18n/en/exam.json (hunk 6 above)
Comment (nicolasruscher) [reply src/main/webapp/i18n/en/exam.json]: done!

File: src/main/webapp/i18n/de/exam.json (hunk 5 above)
Comment (nicolasruscher) [reply src/main/webapp/i18n/de/exam.json]: done!

File: src/main/webapp/app/exam/manage/exam-management.service.ts (hunk 4 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exam/manage/exam-management.service.ts]: done!

File: src/main/webapp/app/exam/manage/students/exam-students.component.ts [hunk 8]
```diff
@@ -233,4 +232,18 @@ export class ExamStudentsComponent implements OnInit, OnDestroy {
         this.jhiAlertService.error(error);
         this.isTransitioning = false;
     }
+
+    /**
+     * Registers all students who are enrolled in the course for the exam
+     */
+    registerAllStudentsFromCourse() {
+        if (this.exam && this.exam.id) {
```
Comment (derLalla) [inline src/main/webapp/app/exam/manage/students/exam-students.component.ts:240]: is it possible for an exam to have the id 0? Not right?
```suggestion
        if (this.exam?.id) {
```

File: src/main/webapp/i18n/de/exam.json [hunk 9]
```diff
@@ -247,6 +247,8 @@
                 "lastName": "Nachname",
                 "email": "E-Mail",
                 "registrationNumber": "Matrikelnummer",
+                "registerAllFromCourse": "Registriere Kurs Studierende",
+                "registerAllFromCourseTooltip": "Registriere alle zu diesem Kurs angemeldeten Studierende",
```
Comment (derLalla) [inline src/main/webapp/i18n/de/exam.json:251]:
```suggestion
                "registerAllFromCourseTooltip": "Registriere alle zu diesem Kurs angemeldeten Studenten",
```

**Review by derLalla — CHANGES_REQUESTED** [review]
besides some minor changes, the code looks good! 👍

**Review by philgabr — DISMISSED** [review]
Works as described, issue that instructors are added as well is resolved :+1:

File: src/main/webapp/app/exam/manage/students/exam-students.component.ts (hunk 8 above)
Comment (nicolasruscher) [reply src/main/webapp/app/exam/manage/students/exam-students.component.ts:240]: good catch!

**Review by philgabr — APPROVED** [review]
Tested new build (11) - re-approve.

**Review by tiffca — APPROVED** [review]
Tested successfully

**Review by valeriebucher — APPROVED** [review]
Works as described.

**Review by DenizShabani — APPROVED** [review]
Looks good

**Review by krusche — APPROVED** [review]

**Pull request merged**